### ISSUE TITLE
operator char*

### ISSUE THREAD

**Issue by quadrustech** [adamyaxley/Obfuscate#43]
Is it probably better to replace the operator "operator char*()" with operator const char*() ?

**Comment by adamyaxley** [comment]
Is there a particular use case that you have in mind where it would be better?

**Comment by geijnmh** [comment]
I am not the OP, but I'd think the class obfuscated_data holds a private member m_data. The operator() automatically decrypts this and returns a pointer to this data for the caller to further process.
I suppose out of good practice one would not want to allow caller to manipulate this data?

If I turn around the question: is there a use case in which caller would want to adjust the data directly?

**Comment by MarioLiebisch** [comment]
I imagine there could be a case where you only want to obfuscate the initial value on disk, but change it at runtime. With this function's return value not being `const`, you allow the obfuscated strings to replace/obfuscate both `const char *some_secret = "sesame";` as well as `char *some_secret = "sesame";`.

**Comment by geijnmh** [comment]
Hi Mario,

Would it not be a segfault waiting to happen because the caller changes the "some_secret" variable into a longer string than was allocated during de-obfuscation? (m_data[N], with N known at compile-time)

**Comment by MarioLiebisch** [comment]
Yes, but you've got that very same issue when just using regular `char[]`, too.

Even a `operator const char*[]()` will expose a similar issue for read access alone.

**Comment by geijnmh** [comment]
Hi,
thanks for sharing your thoughts. I agree that there is a limit how far we should "protect" callers against their own mistakes. But this principle can be applied to the whole encapsulation as well. e.g. why bother making m_data private and obfuscated_data a class, if we want to expose all the risks to the caller.. we could make it a struct with everything public and be done with it).

**Comment by MarioLiebisch** [comment]
`m_data` is private simply to automatically call decryption once you try to retrieve/use the string (again to make it kind of swappable/plug and play).

You can obviously still screw this up, e.g. by retrieving the pointer, then re-encrypting the data and after this possibly modifying the encrypted data, destroying everything.

You can do all this with a struct or raw standalone pointer, but that's not the idea behind C++. This comparison is a bit like using `const char *` vs. `const std::string` for example.

**Comment by geijnmh** [comment]
Well, we could start a lengthy debate about the philosophy of C++, but that would require quotations from Bjarne Strostrup, Herb Sutter, Scott Myers and a few other great names, and would go beyond the topic now, so let us save that for another day.

I disagree with your argument why `m_data` would be private because of automatic decryption. This in fact does not require member privacy at all and this functionality is completely unrelated.
The analogy with `const char*` vs `std::string` does not compute either in my book; the use of `const char*` has a good reason for existence in C++, it depends on the context of the code.

Whether or not `data()` returns `const char*` or `char*` remains a matter of taste and encapsulation design. People who value the design principle of ownership of the class `obfuscated_data` and its responsibility of the memory management of its members, would rather choose `const char*`. Callers can do with the result as they please as long as they copy it into a piece of memory they own.
Those who value caller's responsibility for memory management, even for those items that are allocated by the class would say "Fine. just document this and return `char*`. 

Both are possible and it is for the project maintainer to form an opinion about and I for myself will respect that.

**Comment by adamyaxley** [comment]
Great discussion here.

The main reason why `data()` returns a `char*` instead of `const char*` is to adhere to callers who want to or _need_ to store the string in a `char*`. I agree that returning `const char*` is safer, but it would limit integration into codebases that are using `char*`.

Also, changing `data()` to return a `const char*` at this point could potentially break compatibility with users who are using `char*`, so I'm not keen to do that.

`m_data` is private to make sure that access to it is always done through `data()` and `decrypt()` is always called.

I don't think there is anything that needs to be done now, so I'll close, but always feel free to create a pull request if you have any ideas on how to improve the code.
